distributed/worker.py

```diff
@@ -1496,26 +1505,45 @@ def add_task(
             for dependency, workers in who_has.items():
                 assert workers
                 if dependency not in self.tasks:
+                    # initial state is "new"
+                    # this dependency does not already exist on worker
                     self.tasks[dependency] = dep_ts = TaskState(key=dependency)
-                    dep_ts.state = (
-                        "waiting" if dependency not in self.data else "memory"
-                    )
 
-                dep_ts = self.tasks[dependency]
-                self.log.append((dependency, "new-dep", dep_ts.state))
+                    # link up to child / parents
+                    ts.dependencies.add(dep_ts)
+                    dep_ts.dependents.add(ts)
 
-                if dep_ts.state != "memory":
-                    ts.waiting_for_data.add(dep_ts.key)
-                    self.waiting_for_data_count += 1
+                    # check to ensure task wasn't already executed and partially released
+                    # # TODO: make this less bad
+                    state = "fetch" if dependency not in self.data else "memory"
 
-                dep_ts.who_has.update(workers)
+                    # transition from new -> fetch handles adding dependency
+                    # to waiting_for_data
+                    self.transition(dep_ts, state)
 
-                ts.dependencies.add(dep_ts)
-                dep_ts.dependents.add(ts)
+                    self.log.append((dependency, "new-dep", dep_ts.state, ts))
 
-                for worker in workers:
-                    self.has_what[worker].add(dep_ts.key)
-                    if dep_ts.state != "memory":
+                else:
+                    # task was already present on worker
+                    dep_ts = self.tasks[dependency]
+
+                    # link up to child / parents
+                    ts.dependencies.add(dep_ts)
+                    dep_ts.dependents.add(ts)
+
+                    # possible that a new task has asked for a dependency we're already
+                    # planning on fetching -- make sure that task knows it has to
+                    # wait for its dependencies to arrive
+                    if dep_ts.state in ("fetch", "flight"):
+                        ts.waiting_for_data.add(dep_ts.key)
+
+                if dep_ts.state in ("fetch", "flight"):
```

Comment:
The same condition is checked in the body of the else clause above. I'm wondering if this cannot be fused

Reply:
I think it could, but not if we add `waiting` and `executing` to the check above (which I think is the right things to do)

Reply:
fused